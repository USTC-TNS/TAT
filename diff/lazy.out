1+1=2, 1+2=3
10+11=21
