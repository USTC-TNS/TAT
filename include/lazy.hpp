/**
 * \file lazy.hpp
 *
 * Copyright (C) 2020-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#pragma once
#ifndef LAZY_HPP
#define LAZY_HPP

#ifndef __cplusplus
#error only work for c++
#endif

#if __cplusplus < 201703L
#error require c++17 or later
#endif

#include <functional>
#include <list>
#include <memory>
#include <optional>
#include <stdexcept>
#include <tuple>
#include <type_traits>
#include <unordered_map>
#include <variant>

namespace lazy {
   namespace detail {
      template<class>
      constexpr bool is_reference_wrapper_v = false;
      template<class U>
      constexpr bool is_reference_wrapper_v<std::reference_wrapper<U>> = true;

      template<typename T>
      struct node;

      template<class>
      constexpr bool is_node_v = false;
      template<class U>
      constexpr bool is_node_v<node<U>> = true;

      template<typename Arg>
      const auto& unwrap_arg(const Arg& arg) {
         // arg maybe const X&, std::reference_wrapper or node
         if constexpr (is_node_v<Arg>) {
            return arg();
         } else if constexpr (is_reference_wrapper_v<Arg>) {
            return arg.get();
         } else {
            return arg;
         }
      }

      template<typename Func, typename Args, std::size_t... Is>
      auto unwrap_apply(const Func& func, const Args& args, std::index_sequence<Is...>) {
         return func(unwrap_arg(std::get<Is>(args))...);
      }

      struct general_base : std::enable_shared_from_this<general_base> {
         mutable std::list<std::weak_ptr<general_base>> downstream;
         bool reset() const {
            if (unset()) {
               for (auto i = downstream.begin(); i != downstream.end();) {
                  if (auto ptr = i->lock(); ptr) {
                     ptr->reset();
                     ++i;
                  } else {
                     i = downstream.erase(i);
                  }
               }
               return true;
            } else {
               return false;
            }
         }

         virtual bool unset() const = 0; // true if value changed, false if value not changed
      };

      using map_t = std::unordered_map<std::shared_ptr<general_base>, std::shared_ptr<general_base>>;

      template<typename T>
      struct specific_type : general_base {
         virtual bool reset(const T&) const = 0;
         virtual bool reset(T&&) const = 0;
         virtual bool reset(std::reference_wrapper<const T> u) const = 0;
         virtual const T& get() const = 0;
         virtual node<T> copy(const map_t& map) const = 0;
      };

      template<typename T>
      struct given_value final : specific_type<T> {
         mutable std::variant<std::monostate, T, std::reference_wrapper<const T>> value;
         bool unset() const override {
            if (std::holds_alternative<std::monostate>(value)) {
               return false;
            } else {
               value = std::monostate();
               return true;
            }
         }
         bool reset(const T& u) const override {
            bool result = general_base::reset();
            value = u;
            return result;
         }
         bool reset(T&& u) const override {
            bool result = general_base::reset();
            value = std::move(u);
            return result;
         }
         bool reset(std::reference_wrapper<const T> u) const override {
            bool result = general_base::reset();
            value = u;
            return result;
         }
         const T& get() const override {
            // if it is given_value, it will always have value
            if (auto v = std::get_if<T>(&value)) {
               return *v;
            } else if (auto v = std::get_if<std::reference_wrapper<const T>>(&value)) {
               return v->get();
            }
            throw std::runtime_error("try to get value from empty Root");
         }
         node<T> copy(const map_t& map) const override;
      };

      template<typename T>
      struct specific_function_base : specific_type<T> {
         mutable std::optional<T> value;
         bool unset() const override {
            if (value.has_value()) {
               value.reset();
               return true;
            } else {
               return false;
            }
         }
         bool reset(const T& u) const override {
            throw std::runtime_error("try to set value of functional node");
         }
         bool reset(T&& u) const override {
            throw std::runtime_error("try to set value of functional node");
         }
         bool reset(std::reference_wrapper<const T> u) const override {
            throw std::runtime_error("try to set value of functional node");
         }
      };

      template<typename T, typename Func, typename... Args>
      struct specific_function final : specific_function_base<T> {
         using specific_function_base<T>::value;
         Func func;                // it can be called directly if Func is std::reference_wrapper
         std::tuple<Args...> args; // if it needs reference, use std::cref
         template<typename Func0, typename... Args0>
         specific_function(Func0&& func0, Args0&&... args0) : func(std::forward<Func0>(func0)), args(std::forward<Args0>(args0)...) {}
         const T& get() const override {
            // if it is given_value, it will always have value
            if (!value.has_value()) {
               value = unwrap_apply(func, args, std::index_sequence_for<Args...>());
            }
            return value.value();
         }
         node<T> copy(const map_t& map) const override;
      };

      template<typename T>
      struct node {
         using pointer_t = std::shared_ptr<specific_type<T>>;
         pointer_t pointer;
         node(pointer_t&& p) : pointer(std::move(p)) {}
         bool reset() const {
            return pointer->reset();
         };
         bool reset(const T& u) const {
            return pointer->reset(u);
         };
         bool reset(T&& u) const {
            return pointer->reset(std::move(u));
         };
         bool reset(std::reference_wrapper<const T> u) const {
            return pointer->reset(u);
         };
         const T& operator()() const {
            return pointer->get();
         }
      };

      template<typename Down, typename Up>
      bool try_add_downstream(const Down& down, const Up& up) {
         // down is shared_ptr of specific_function
         // up is node or normal object
         if constexpr (is_node_v<Up>) {
            up.pointer->downstream.push_back(down);
            return true;
         } else {
            return false;
         }
      }

      template<typename T, std::size_t... Is>
      void add_as_downstream(const T& t, std::index_sequence<Is...>) {
         // T is a shared_ptr of specific_function
         (try_add_downstream(t, std::get<Is>(t->args)), ...);
      }

      template<typename Func, typename... Args>
      auto Node(Func&& func, Args&&... args) {
         using T = std::invoke_result_t<Func, decltype(unwrap_arg(std::declval<Args>()))...>;
         auto result = std::make_shared<specific_function<T, std::remove_reference_t<Func>, std::remove_reference_t<Args>...>>(
               std::forward<Func>(func),
               std::forward<Args>(args)...);
         add_as_downstream(result, std::index_sequence_for<Args...>());
         return node<T>(std::move(result));
      }
      template<typename T>
      auto Root() {
         auto result = std::make_shared<given_value<T>>();
         return node<T>(std::move(result));
      }
      template<typename T>
      auto Root(const T& arg) {
         auto result = std::make_shared<given_value<T>>();
         result->reset(arg);
         return node<T>(std::move(result));
      }
      template<typename T>
      auto Root(T&& arg) {
         auto result = std::make_shared<given_value<T>>();
         result->reset(std::move(arg));
         return node<T>(std::move(result));
      }
      template<typename T>
      auto Root(std::reference_wrapper<const T> arg) {
         auto result = std::make_shared<given_value<T>>();
         result->reset(arg);
         return node<T>(std::move(result));
      }

      struct Copy {
         map_t map;
         template<typename T>
         node<T> operator()(const node<T>& n0) {
            node<T> result = n0.pointer->copy(map);
            map[n0.pointer] = result.pointer;
            return result;
         }
      };

      template<typename T>
      node<T> given_value<T>::copy(const map_t& map) const {
         auto result = std::make_shared<given_value<T>>();
         if (auto v = std::get_if<T>(&value)) {
            result->reset(*v); // copy value
         } else if (auto v = std::get_if<std::reference_wrapper<const T>>(&value)) {
            result->reset(*v); // copy reference_wrapper
         }
         return node<T>(std::move(result));
      }

      template<typename T>
      decltype(auto) try_map(const map_t& map, const T& arg) {
         if constexpr (is_node_v<T>) {
            // T is node<U>
            return T(std::dynamic_pointer_cast<typename T::pointer_t::element_type>(map.at(arg.pointer))); // value
         } else {
            return (arg); // reference
         }
      }

      template<typename T, typename Func, typename Args, std::size_t... Is>
      auto copy_specific_function_impl(const Func& func, const Args& args, const map_t& map, std::index_sequence<Is...>) {
         return std::make_shared<T>(func, try_map(map, std::get<Is>(args))...);
      }

      template<typename T, typename Func, typename... Args>
      node<T> specific_function<T, Func, Args...>::copy(const map_t& map) const {
         auto result = copy_specific_function_impl<specific_function<T, Func, Args...>>(func, args, map, std::index_sequence_for<Args...>());
         add_as_downstream(result, std::index_sequence_for<Args...>());
         result->value = value;
         return node<T>(std::move(result));
      }
   } // namespace detail

   using detail::Copy;
   using detail::Node;
   using detail::Root;
} // namespace lazy

#endif
