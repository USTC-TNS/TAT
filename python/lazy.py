#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (C) 2020-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

from __future__ import annotations
from typing import TypeVar, Generic
import weakref
from multimethod import multimethod

__all__ = ["Copier", "Node", "Root"]


class Copier:

    def __init__(self):
        self._map = {}

    def __call__(self, node):
        result = Node(node, self)
        self._map[node] = result
        return result

    def _map_node(self, node):
        if isinstance(node, Node):
            return self._map[node]
        else:
            return node


T = TypeVar('T')


class Node(Generic[T]):

    @multimethod
    def __init__(self, other: Node[T], copier: Copier):
        self._value: T | None = other._value
        self._downstream: set[Node] = set()
        self._func = other._func
        self._args = [copier._map_node(i) for i in other._args]
        self._kwargs = {i: copier._map_node(j) for i, j in other._kwargs.items()}

        for i in self._args:
            if isinstance(i, Node):
                i._downstream.add(weakref.ref(self))

    @multimethod
    def __init__(self, func, *args, **kwargs):
        self._value: T | None = None
        self._downstream: set[Node] = set()
        self._func = func
        self._args = args
        self._kwargs = kwargs

        for i in self._args:
            if isinstance(i, Node):
                i._downstream.add(weakref.ref(self))

    def reset(self, value: T | None = None):
        if self._value != value:
            self._value = value
            new_downstream: set[Node] = set()
            for i in self._downstream:
                if i():
                    i().reset()
                    new_downstream.add(i)

            self._downstream = new_downstream

    def __call__(self) -> T:
        if self._value is None:
            args = [self._unwrap_node(i) for i in self._args]
            kwargs = {i: self._unwrap_node(j) for i, j in self._kwargs.items()}
            self._value = self._func(*args, **kwargs)
        return self._value

    @staticmethod
    def _unwrap_node(node):
        if isinstance(node, Node):
            return node()
        else:
            return node


def Root(value: T | None = None) -> Node[T]:
    result = Node(lambda: None)
    result.reset(value)
    return result
