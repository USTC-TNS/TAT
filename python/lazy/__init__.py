#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (C) 2020-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#

from __future__ import annotations
from typing import TypeVar, Generic, Callable
import weakref

__all__ = ["Copier", "Merger", "Node", "Root"]


class Copier:
    __slots__ = ["_map"]

    def __init__(self) -> None:
        self._map: dict[Node, Node] = {}

    def __call__(self, node: Node) -> Node:
        result: Node = Node._copy(node, self)
        result._value = node._value
        self._map[node] = result
        return result

    def _map_node(self, node):
        if isinstance(node, Node):
            return self._map[node]
        else:
            return node


class Merger:
    __slots__ = ["_map", "_select_1", "_select_2"]

    def __init__(self) -> None:
        self._map: dict[Node, Node] = {}
        self._select_1: dict[Node] = set()  # nodes using value from channel 1
        self._select_2: dict[Node] = set()  # nodes using value from channel 2

    def __call__(self, node_1: Node, node_2: Node, select: int) -> Node:
        if select == 1:
            result: Node = Node._copy(node_1, self)
            result._value = node_1._value
            self._select_1(result)
        elif select == 2:
            result: Node = Node._copy(node_2, self)
            result._value = node_2._value
            self._select_2(result)
        elif select == 0:
            result: Node = Node._copy(node_1, self)
            select_1: bool = True
            select_2: bool = True
            upstreams: set[Node] = {i for i in result._args if isinstance(i, Node)} | {j for i, j in result._kwargs.items() if isinstance(j, Node)}
            for i in upstreams:
                if i not in self._select_1:
                    select_1 = False  # upstream not using channel 1 so result cannot use it
                if i not in self._select_2:
                    select_2 = False  # upstream not using channel 2 so result cannot use it
            if select_1:
                result._value = node_1._value
                self._select_1.add(result)
            elif select_2:
                result._value = node_2._value
                self._select_2.add(result)
            else:
                result._value = None
        else:
            raise ValueError("Invalid select index")
        self._map[node_1] = result
        self._map[node_2] = result
        return result

    def _map_node(self, node):
        if isinstance(node, Node):
            return self._map[node]
        else:
            return node


T = TypeVar('T')


class Node(Generic[T]):
    __slots__ = ["_value", "_downstream", "_func", "_args", "_kwargs", "__weakref__"]

    @staticmethod
    def _copy(other: Node[T], copier: Copier | Merger) -> Node[T]:
        args = [copier._map_node(i) for i in other._args]
        kwargs = {i: copier._map_node(j) for i, j in other._kwargs.items()}
        return Node(other._func, *args, **kwargs)

    def __init__(self, func: Callable[..., T], *args, **kwargs):
        self._value: T | None = None
        self._downstream: set[weakref.ref[Node]] = set()
        self._func: [..., T] = func
        self._args = args
        self._kwargs = kwargs

        for i in self._args:
            if isinstance(i, Node):
                i._downstream.add(weakref.ref(self))

    def reset(self, value: T | None = None):
        if self._value != value:
            self._value = value
            new_downstream: set[weakref.ref[Node]] = set()
            for i in self._downstream:
                if i():
                    i().reset()
                    new_downstream.add(i)

            self._downstream = new_downstream

    def __call__(self) -> T:
        if self._value is None:
            # need to reduce stack depth
            # or call sys.setrecursionlimit(more_stack_depth)

            # args = [self._unwrap_node(i) for i in self._args]
            args = []
            for i in self._args:
                if isinstance(i, Node):
                    i = i()
                args.append(i)

            # kwargs = {i: self._unwrap_node(j) for i, j in self._kwargs.items()}
            kwargs = {}
            for i, j in self._kwargs.items():
                if isinstance(j, Node):
                    j = j()
                kwargs[i] = j

            self._value = self._func(*args, **kwargs)
        return self._value


def Root(value: T | None = None) -> Node[T]:
    result = Node(lambda: None)
    result.reset(value)
    return result
