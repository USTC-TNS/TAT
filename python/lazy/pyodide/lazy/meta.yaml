package:
  name: lazy
  version: 0.1.4

source:
  url: https://github.com/hzhangxyz/TAT/archive/refs/heads/feature/pyodide.zip
  extras:
    - [setup.py, setup.py]
