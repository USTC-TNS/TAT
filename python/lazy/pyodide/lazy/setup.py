from distutils.core import setup

setup(name="lazy", version="0.1.4", packages=["lazy"], package_dir={"": "python"})
