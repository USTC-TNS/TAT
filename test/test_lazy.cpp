/**
 * Copyright (C) 2020-2021 Hao Zhang<zh970205@mail.ustc.edu.cn>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <iostream>
#include <lazy.hpp>

#include "run_test.hpp"

void run_test() {
   {
      auto v = 1;
      auto a = lazy::Root(std::cref(v));
      auto b = lazy::Root(2);
      std::cout << a() << "\n";
      std::cout << b() << "\n";
      auto c = lazy::Node(
            [](int a, int b) {
               return a + b;
            },
            a,
            b);
      auto d = lazy::Node(
            [](int c, int a) {
               return c * a;
            },
            c,
            a);
      std::cout << d() << "\n";
      a.reset(233);
      std::cout << d() << "\n";
      b.reset(666);
      std::cout << d() << "\n";
   }
   {
      auto a = lazy::Root(1);
      auto b = lazy::Root(10);
      auto c = lazy::Node(
            [](auto i, auto j) {
               return i + j;
            },
            a,
            b);
      std::cout << c() << "\n";
      b.reset(100);
      std::cout << c() << "\n";
      auto copy = lazy::Copy();
      auto aa = copy(a);
      auto bb = copy(b);
      auto cc = copy(c);
      aa.reset(1000);
      std::cout << cc() << "\n";
      std::cout << c() << "\n";
   }
}
