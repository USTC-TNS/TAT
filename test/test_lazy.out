1
2
3
54755
209467
11
101
1100
101
